### ISSUE TITLE
Inserted text from Clipboard as plaintext

### ISSUE THREAD

**Issue by confile** [Voog/wysihtml#195]
How can I insert text from Clipboard as plaintext without any format?

**Comment by robbytaylor** [comment]
The easiest way would be to not use Wysihtml at all and and use contenteditable. You can then catch the paste event on the editable element to ensure that the pasted content doesn't contain HTML.

See [this stackoverflow answer](http://stackoverflow.com/questions/12027137/javascript-trick-for-paste-as-plain-text-in-execcommand/12028136#12028136) about intercepting pasted content.

You'd need to do a bit more to ensure that the pasted content doesn't contain HTML. With jQuery you could extend the above to:

```
editor.addEventListener("paste", function(e) {
    // cancel paste
    e.preventDefault();

    // get text representation of clipboard
    var html = e.clipboardData.getData("text/plain"),
        text = $('<div>' + html '</div>').text();

    // insert text manually
    document.execCommand("insertHTML", false, text);
});
```

If you're stripping HTML then you might also want to go to the trouble of removing new lines (depending on what you're building) in which case you can add:

```
text = text.replace(/\n|\r|\n\r/g, '');
```

**Comment by confile** [comment]
@robbytaylor Thanks for the answer but is there also a way of doing it with wysihtml?

My problem is when I copy and paste to wysihtml I often get content like this: 

```
<h1><span style="some inline style">Text</span></h1>
```

This kills my css style for h1. 

Is there any way to prevent this? What would you do in this case? Thanks for your help.

**Comment by robbytaylor** [comment]
You could set all the HTML tags to be unwrapped in the parser rules but that might be a bit of a faff.

It looks like your issue might be in your parser rules anyway as the inline styles should have been removed. Are you using custom parser rules or one of the defaults? Are you able to show the code you're using?

**Comment by confile** [comment]
Sure here is my code. I took it from the example: 

```
		var editor = new wysihtml5.Editor('editor', {
		    toolbar: 'toolbar',
		    parserRules:  wysihtml5ParserRules
		});
		editor
	    .on("load", function() {
		  // show editor only after rendering is done
	      $("#editorWrapper").show();
	    });
```

Is there anything you would to change. Escaping the inline style would be great.

**Comment by robbytaylor** [comment]
That could be a bug. I just tried on the example at http://wysihtml.com/ and got a similar - I was able to paste an element with an inline style.

I don't know whether it's definitely a bug or the expected behaviour for the rules wysihtml.com is using but the same doesn't happen with the simple example at http://voog.github.io/wysihtml/examples/simple.html

The parser rules that demo is using are at https://github.com/Voog/wysihtml/blob/master/parser_rules/simple.js so you could try using those instead. If you've taken the code from the example there might be a file called parserrules.js which could can replace.

**Comment by confile** [comment]
What is the difference between the simple and the advanced rule and the parserrules.js?

**Comment by robbytaylor** [comment]
It should just change which tags and classes are allowed, I'm clutching at straws now really!

It looks like the example on wysihtml.com is using version 0.5.0-beta2 so I think this must be a bug, which has possibly been fixed already as the latest version of the editor is 0.5.0-beta11. You could try upgrading to 0.5.0-beta11 or downgrading to the latest stable version (0.4.17).

**Comment by confile** [comment]
@robbytaylor Which parser rule would you recommend?

I used ``advanced_and_extended.js``before which leads to this problem.

**Comment by pulges** [comment]
The easiest way if using advanced parser rules (that have special pasted cleanup rule-sets in addition to some other things) is to change the wysihtml5ParserPasteRulesets object so the rules set is empty on paste. This way no html markup should be kept on paste.

    window.wysihtml5ParserPasteRulesets = [
        {
            set: {}
        }
    ];

You can remove other functions in this scope too as they are not used in this case.

If you want formatting to be kept on copy-paste from wysihtml editable area to the wysihtml editable area, and be cleaned if from other sources, you can use something like this:

    window.wysihtml5ParserPasteRulesets = [
        {
            condition: /<meta name="copied-from" content="wysihtml5">/i,
            set: commonRules
        },{
            set: {}
        }
    ];

**Comment by confile** [comment]
Thank you very much!

**Comment by confile** [comment]
@pulges Isn't the simple parser rule sufficient?

**Comment by pulges** [comment]
It all depends what you want to achieve. The advanced and extended is the maximal functionality showing case where a lot is not needed for most use cases.
